src/robot.coffee

```diff
@@ -503,4 +497,80 @@ class Robot
     HttpClient.create(url, options)
       .header('User-Agent', "Hubot/#{@version}")
 
+  # Public. Adds a prereceive callback hook to this robot.
+  #
+  #   cb  - A callback function which accepts a Hook object.
+  #
+  # See also Hook
+  # Returns nothing
+  prereceive: (cb) ->
+    @hooks['prereceive'] ||= []
+    @hooks['prereceive'].push cb
+
+  # Public. Adds a prelisten callback hook to this robot.
+  #
+  #   cb  - A callback function which accepts a Hook object.
+  #
+  # See also Hook
+  # Returns nothing
+  prelisten: (cb) ->
+    @hooks['prelisten'] ||= []
+    @hooks['prelisten'].push cb
+
+  # Protected. For use by Listener to run prelisten callbacks after a match
+  # is found.
+  #
+  # See also Hook
+  # Returns nothing
+  runPrelistenHooks: (listener, response) ->
+    done = ->
+      listener.callback(response) unless response.envelope.message.done
+    @runHooks 'prelisten', done, response.envelope.message, response, listener
+
+  # Public. Adds a prereply callback hook to this robot.
+  #
+  #   cb  - A callback function which accepts a Hook object.
+  #
+  # See also Hook
+  # Returns nothing
+  prereply: (cb) ->
+    @hooks['prereply'] ||= []
+    @hooks['prereply'].push cb
+
+  # Protected. For use by Response to run hooks before replying.
+  #
+  # TODO: we should have a listener available here, but there's no way to get
+  # to it. Responses belong to listeners but don't have a way to get to them,
+  # and we should change that API.
+  runPrereplyHooks: (response, string, cb) ->
+    @runHooks 'prereply', cb, response.envelope.message, response, response.listener, { text: string }
+
+  # Private. Run the hooks of a given name
+  #
+  #  name     - 'prereceive'
+  #  response - The response object
+  #  listener - The matching listener, if any
+  #  message  - A message that hubot is trying to send, if any
+  #
+  # Each hook must call hook.next() or hook.done(). Hook.next() continues
+  # processing and hook.done() aborts the message. If a response message
+  # exists, hook.done() prevents it from being sent.
+  #
+  # Returns nothing.
+  runHooks: (name, callback, message, response, listener, reply) ->
```

Comment:
lol. I would like to clean this up by making response aware of messages and listeners, and letting `match` and `listener` potentially be null.